core/services/health/health.go

```diff
@@ -0,0 +1,212 @@
+package health
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+// Types requiring health checks should implement the Checkable interface.
+type Checkable interface {
+	// Checkables should return nil if ready, or an error message otherwise.
+	Ready() error
+	// Checkables should return nil if healthy, or an error message otherwise.
+	Healthy() error
+}
+
+//go:generate mockery --name Checker --output ../../internal/mocks/ --case=underscore
+type (
+	// Checker provides a service which can be probed for system health.
+	Checker interface {
+		// Register a service for health checks.
+		Register(name string, service Checkable) error
+		// Unregister a service.
+		Unregister(name string) error
+		// IsReady returns the current readiness of the system.
+		// A system is considered ready if all checks are passing (no errors)
+		IsReady() (ready bool, errors map[string]error)
+		// IsHealthy returns the current health of the system.
+		// A system is considered healthy if all checks are passing (no errors)
+		IsHealthy() (healthy bool, errors map[string]error)
+
+		Start() error
+		Close() error
+	}
+
+	checker struct {
+		srvMutex   sync.RWMutex
+		services   map[string]Checkable
+		stateMutex sync.RWMutex
+		state      map[string]State
+
+		stop chan struct{}
+		done chan struct{}
+
+		utils.StartStopOnce
+	}
+
+	State struct {
+		ready   error
+		healthy error
+	}
+
+	Status string
+)
+
+var _ Checker = (*checker)(nil)
+
+const (
+	StatusPassing Status = "passing"
+	StatusFailing Status = "failing"
+
+	interval = 15 * time.Second
+)
+
+var (
+	metrics = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "health",
+		Help: "Health status by service",
+	},
+		[]string{"service_id"},
+	)
+)
+
+func NewChecker() Checker {
+	c := &checker{
+		services: make(map[string]Checkable, 10),
+		state:    make(map[string]State, 10),
+		stop:     make(chan struct{}),
+		done:     make(chan struct{}),
```

Comment:
minor, but i think there is already a naming convention for these: chStop, chDone